controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,377 @@
+//      ii) That the infrastructureMachineTemplate has the correct labels and annotations
+func assertControlPlaneReconcile(cluster *clusterv1.Cluster, clusterClass *clusterv1.ClusterClass) bool {
+	cp, ok := getAndAssertLabelsAndAnnotations(*cluster.Spec.ControlPlaneRef, cluster.Name)
+	if !ok {
+		return false
+	}
+	// Check if the ControlPlane Version matches the version in the Cluster's managed topology spec.
+	version, err := contract.ControlPlane().Version().Get(cp)
+	if err != nil {
+		log.Infof("We failed on getting control plane version %v %v", err, cp)
+		return false
+	}
+	if *version != cluster.Spec.Topology.Version {
+		log.Infof("We failed on comparing control plane version %v", cp)
+		return false
+	}
+	// Check for Control Plane replicase if it's set in the Cluster.Spec.Topology
+	if cluster.Spec.Topology.ControlPlane.Replicas != nil {
+		replicas, err := contract.ControlPlane().Replicas().Get(cp)
+		if err != nil {
+			log.Infof("We failed on getting control plane replicas %v", cp)
+			return false
+		}
+		// Check for Control Plane replicase if it's set in the Cluster.Spec.Topology
+		if int32(*replicas) != *cluster.Spec.Topology.ControlPlane.Replicas {
+			log.Infof("We failed on comparing control plane replicas %v", cp)
+			return false
+		}
+	}
+	log.Infof("ControlPlane passed with %v", cp)
+	if clusterClass.Spec.ControlPlane.MachineInfrastructure != nil && clusterClass.Spec.ControlPlane.MachineInfrastructure.Ref != nil {
+		cpInfra, err := contract.ControlPlane().MachineTemplate().InfrastructureRef().Get(cp)
+		if err != nil {
+			log.Infof("We failed on getting control plane infrastructure %v,  %v", err, cp)
+			return false
+		}
+		if ok := referenceExistsWithCorrectKindAndAPIVersion(cpInfra,
+			builder.GenericInfrastructureMachineTemplateKind,
+			builder.InfrastructureGroupVersion); !ok {
+			log.Infof("Kind or version wrong from InfrastructureMachineTemplate %v", cp)
+			return false
+		}
+		if _, ok := getAndAssertLabelsAndAnnotations(*cpInfra, cluster.Name); !ok {
+			return false
+		}
+		log.Infof("ControlPlaneInfrastructure passed with %v", cpInfra)
+	}
+	return true
+}
+
+// assertMachineDeploymentsReconcile checks if the MachineDeployments:
+// 1) Are created in the correct number.
+// 2) Have the correct labels (TopologyOwned, ClusterName, MachineDeploymentName).
+// 3) Have the correct finalizer applied.
+// 4) Have the correct replicas and version.
+// 6) Have the correct Kind/APIVersion and Labels/Annotations for BoostrapRef and InfrastructureRef templates.
+func assertMachineDeploymentsReconcile(cluster *clusterv1.Cluster) bool {
+	// List all created machine deployments to assert the expected numbers are c*reated.
+	machineDeployments := &clusterv1.MachineDeploymentList{}
+	if err := env.List(ctx, machineDeployments); err != nil {
+		log.Infof("machine deployments not listed %v", err)
+		return false
+	}
+	clusterMDs := make([]clusterv1.MachineDeployment, 0)
```

Comment:
nit: if we explicitly create the slice with make should we use the correct size? `len(machineDeployments.Items)`

(I'm personally fine with just `var clusterMDs []clusterv1.MachineDeployment`)

Reply:
I'll change it to the var - putting in the expected size seems confusing with no upside, but I might be missing something there.

Reply:
As far as I'm aware the make syntax is usually only used if the slice has to be created with a specific size (usually for performance reasons, to avoid copying the array on append). So that's why I just was a bit confused why we're using the make syntax to create an empty slice.

Reply:
In case this gives a linter warning I'll update my recommendation to `clusterMDs := []clusterv1.MachineDeployment{}` :)

Reply:
>In case this gives a linter warning I'll update my recommendation to clusterMDs := []clusterv1.MachineDeployment{} :)

There's a linter for that too 😆  But only locally - it'll pass make lint.